### ISSUE TITLE
'force-command' ignored for PTY sessions

### ISSUE THREAD

**Issue by alukyan** [PowerShell/Win32-OpenSSH#1408]
**"OpenSSH for Windows" version**
7.9.0.0
 
**Server OperatingSystem**
Windows Server 2016 Datacenter

**Client OperatingSystem**
Ubuntu 18.04

**What is failing**
I am using authentication with SSH certificates. And I created a certificate that contains 'force-command' option

`ssh-keygen -s .ssh/ca_userkeys -I cert_identity -V -1w:+54w5d -n myuser -O force-command='cmd.exe /c "dir C:"' .ssh/id_user.pub`

```
ssh-keygen -Lf .ssh/id_user-cert.pub
.ssh/id_user-cert.pub:
        Type: ssh-rsa-cert-v01@openssh.com user certificate
        Public key: RSA-CERT SHA256:gGDks104o/hFOoWVgFGYzea0tIvE2rKrH6mYXdFowJg
        Signing CA: RSA SHA256:e3yqj7Cb6BkqiRw5YCAy2oRy3rfnIg8DQNWXROkfgyE
        Key ID: "cert_identity"
        Serial: 0
        Valid: from 2019-06-14T16:37:24 to 2020-07-08T16:37:24
        Principals: 
                myuser
        Critical Options: 
                force-command cmd.exe /c "dir C:"
        Extensions: 
                permit-X11-forwarding
                permit-agent-forwarding
                permit-port-forwarding
                permit-pty
                permit-user-rc

```

**Expected output**
It is expected that after connection the `force-command` will be executed and the session will end. 

**Actual output**
Regular Interactive connection is established and `force-command` is ignored.

**Comment by manojampalam** [comment]
Thanks for catching this. The issue applies to ForceCommand with PTY sessions in general (and not restricted to Certificates alone). 

This is due to a bug in PTY sessions handler code for Windows - it seems to ignore user/session passed commandline - https://github.com/PowerShell/openssh-portable/blob/latestw_all/contrib/win32/win32compat/w32-doexec.c#L319
Updated wiki on how to workaround:
https://github.com/PowerShell/Win32-OpenSSH/wiki/sshd_config#forcecommand

**Comment by alukyan** [comment]
Workaround with `no-pty` does not seem to work.

I regenerated the certificate
```
~$ ssh-keygen -s .ssh/ca_userkeys -I cert_identity -V -1w:+54w5d -n myuser -O force-command='cmd.exe /c "dir C:"' -O no-pty .ssh/id_user.pub
Signed user key .ssh/id_user-cert.pub: id "cert_identity" serial 0 for myuser valid from 2019-06-18T18:01:51 to 2020-07-12T18:01:51

~$ ssh-keygen -Lf .ssh/id_user-cert.pub 
.ssh/id_user-cert.pub:
        Type: ssh-rsa-cert-v01@openssh.com user certificate
        Public key: RSA-CERT SHA256:gGDks104o/hFOoWVgFGYzea0tIvE2rKrH6mYXdFowJg
        Signing CA: RSA SHA256:e3yqj7Cb6BkqiRw5YCAy2oRy3rfnIg8DQNWXROkfgyE
        Key ID: "cert_identity"
        Serial: 0
        Valid: from 2019-06-18T18:01:51 to 2020-07-12T18:01:51
        Principals: 
                myuser
        Critical Options: 
                force-command cmd.exe /c "dir C:"
        Extensions: 
                permit-X11-forwarding
                permit-agent-forwarding
                permit-port-forwarding
                permit-user-rc
```
`force-command` still does not work as it works on Linux. It connects as usual and shows the interactive prompt. 
I see no output from `cmd.exe /c "dir C:"`

**Comment by manojampalam** [comment]
Just tried and it seems to work:
```
c:\> ssh-keygen -s ca_key -I 1 -V -1w:+54w5d -n user -O force-command="cmd.exe /c echo force-command" -O no-pty  k1.pub
c:\> ssh -i .\k1 user@localhost
PTY allocation request failed on channel 0
force-command
Connection to localhost closed.
```

If you see otherwise, share server side traces.

**Comment by manojampalam** [comment]
Tagging test coverage to add a test case to cover above scenario.

**Comment by alukyan** [comment]
I see you are connecting to local box. I am trying to sign keys on Windows VM and I hit some bug with file access:
```
C:\Users\griffink8s\.ssh>dir
 Volume in drive C is Windows
 Volume Serial Number is 8E44-3BE0

 Directory of C:\Users\griffink8s\.ssh

06/27/2019  07:27 PM    <DIR>          .
06/27/2019  07:27 PM    <DIR>          ..
06/25/2019  11:59 PM               796 authorized_keys
06/27/2019  07:22 PM             1,706 ca_userkeys
06/27/2019  07:20 PM               409 ca_userkeys.pub
06/27/2019  07:23 PM             2,610 id_my
06/27/2019  07:23 PM               577 id_my.pub
               5 File(s)          6,098 bytes
               2 Dir(s)  24,835,764,224 bytes free

C:\Users\griffink8s\.ssh>"C:\Program Files\OpenSSH\ssh-keygen.exe" -s ca_userkeys -I cert_identity -V -1w:+54w5d -n griffink8s -O force-command='cmd.exe /c echo anything' -O no-pty id_my.pub
do_ca_sign: unable to open "/c": No such file or directory
```

When I connect from my Linux box the behavior gets even weirder
```
~/.ssh$ ssh-keygen -s ca_userkeys -I cert_identity -V -1w:+54w5d -n griffink8s -O force-command='cmd.exe /c echo force-command' -O no-pty id_my.pub
Signed user key id_my-cert.pub: id "cert_identity" serial 0 for griffink8s valid from 2019-06-20T12:38:59 to 2020-07-14T12:38:59

~/.ssh$ ssh -i id_my-cert.pub griffink8s@13.65.211.32 
PTY allocation request failed on channel 0
force-command
Connection to 13.65.211.32 closed.
```

Now with different force-command
```
~/.ssh$ ssh-keygen -s ca_userkeys -I cert_identity -V -1w:+54w5d -n griffink8s -O force-command='cmd.exe /c echo anything' -O no-pty id_my.pub
Signed user key id_my-cert.pub: id "cert_identity" serial 0 for griffink8s valid from 2019-06-20T12:40:30 to 2020-07-14T12:40:30
~/.ssh$ ssh -i id_my-cert.pub griffink8s@13.65.211.32 
PTY allocation request failed on channel 0
force-command
Connection to 13.65.211.32 closed.
```
**same output** as before!!!
Something is really broken here. How do I get server side traces - I do not see any logs in `C:\ProgramData\ssh\logs`

**Comment by matsmcp** [comment]
Any updates on this Issue?
It makes creating a good jumphost config a bit harder

**Comment by bagajjal** [comment]
@matsmcp - Can you try V8.1.0.0. As per earlier conversation, this seems to work.

**Comment by bagajjal** [comment]
@matsmcp - If v8.1.0.0 doesn't work then try the [private sshd binary](https://github.com/PowerShell/Win32-OpenSSH/files/6213475/sshd_latest.zip). 
Please follow these steps
1) cd to openssh bin folder.
2) stop the sshd service
net stop sshd
2) take backup of sshd.exe
3) copy the private sshd binary to bin folder.
4) start sshd service
net start sshd
5) try your scenario.

**Comment by matsmcp** [comment]
@bagajjal 
I tried 8.1.0.0. Does not work.
The Private fix do work as expected. I hope it gets included in 8.5

**Comment by bagajjal** [comment]
@matsmcp - Great.
